runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/sideinput/SideInputCache.java

```diff
@@ -0,0 +1,106 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow.worker.streaming.sideinput;
+
+import com.google.auto.value.AutoValue;
+import com.google.errorprone.annotations.CheckReturnValue;
+import java.util.Optional;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.cache.Cache;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.cache.CacheBuilder;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.cache.Weigher;
+
+/**
+ * Wrapper around {@code Cache<SideInputId, SideInput>} that mostly delegates to the underlying
+ * cache, but adds threadsafe functionality to invalidate and load entries that are not ready.
+ *
+ * @implNote Returned values are explicitly cast, because the {@link #sideInputCache} holds wildcard
+ *     types of all objects.
+ */
+@CheckReturnValue
+final class SideInputCache {
+
+  private static final long MAXIMUM_CACHE_WEIGHT = 100000000; /* 100 MB */
+  private static final long CACHE_ENTRY_EXPIRY_MINUTES = 1L;
+
+  private final Cache<Key, SideInput<?>> sideInputCache;
+
+  SideInputCache(Cache<Key, SideInput<?>> sideInputCache) {
+    this.sideInputCache = sideInputCache;
+  }
+
+  static SideInputCache create() {
+    return new SideInputCache(
+        CacheBuilder.newBuilder()
+            .maximumWeight(MAXIMUM_CACHE_WEIGHT)
+            .expireAfterWrite(CACHE_ENTRY_EXPIRY_MINUTES, TimeUnit.MINUTES)
+            .weigher((Weigher<Key, SideInput<?>>) (id, entry) -> entry.size())
+            .build());
+  }
+
+  synchronized <T> SideInput<T> invalidateThenLoadNewEntry(
+      Key key, Callable<SideInput<T>> cacheLoaderFn) throws ExecutionException {
+    // Invalidate the existing not-ready entry.  This must be done atomically
+    // so that another thread doesn't replace the entry with a ready entry, which
+    // would then be deleted here.
+    SideInput<?> newEntry = sideInputCache.getIfPresent(key);
+    if (newEntry != null && !newEntry.isReady()) {
+      sideInputCache.invalidate(key);
+    }
+
+    @SuppressWarnings({
+      "unchecked" // cacheLoaderFn loads SideInput<T>, so value for Key is always SideInput<T>.
```

Comment:
the comment here and elsewhere doesn't entirely explain why the cast is safe.
If we're using the cacheLoaderFn from this invocation then it's safe, but if there was a previous call to get for the same key with a different type and that was cached, the cast would be incorrect.

So I think the comment should be that the runner enforces that the same type is used for the same key. (we could perhaps enforce this with some instanceof check or making type part of the cache key?)

Reply:
done, i added the `PCollectionView.viewFn.typeDescriptor` to the cache key.